File: src/codegen/subcollection/subcollection.py
#
# Copyright (c) 2010 Red Hat, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#


from ovirtsdk.utils.parsehelper import ParseHelper
from codegen.utils.typeutil import TypeUtil
from codegen.doc.documentation import Documentation
from codegen.utils.paramutils import ParamUtils
from codegen.utils.urlutils import UrlUtils
from codegen.utils.headerutils import HeaderUtils
from codegen.templates.subcollectiontemplate import SubCollectionTemplate
from codegen.templates.subcollectiongettemplate import SubCollectionGetTemplate
from codegen.templates.subcollectionlisttemplate import SubCollectionListTemplate
from codegen.templates.subcollectionlistwithparamstemplate import SubCollectionListWithParamsTemplate
from codegen.templates.subcollectionaddtemplate import SubCollectionAddTemplate

#============================================================
#======================SUB COLLECTION========================
#============================================================

subcollectiontemplate = SubCollectionTemplate()
subcollectiongettemplate = SubCollectionGetTemplate()
subcollectionlisttemplate = SubCollectionListTemplate()
subcollectionlistwithparamstemplate = SubCollectionListWithParamsTemplate()
subcollectionaddtemplate = SubCollectionAddTemplate()

class SubCollection(object):

    @staticmethod
    def collection(sub_collection_name, parent_resource_name_lc):

        sub_collection_get_template_values = {
          'sub_collection_name'    : sub_collection_name,
          'parent_resource_name_lc': parent_resource_name_lc.lower()
        }

        return subcollectiontemplate\
               .generate(sub_collection_get_template_values)

    @staticmethod
    def get(url, link, parent_resource_name_lc, encapsulating_resource,
            actual_resource_name_candidate, KNOWN_WRAPPER_TYPES={},
            NAMING_ENTITY_EXCEPTIONS={}):

        actual_encapsulating_resource = \
            TypeUtil.getValueByKeyOrNone(
                 encapsulating_resource.lower(),
                 KNOWN_WRAPPER_TYPES
             )

        actual_resource_name_lc = (
           ParseHelper.getXmlTypeInstance(
              actual_encapsulating_resource \
                    if actual_encapsulating_resource is not None \
                    else actual_resource_name_candidate.lower()
              )
           ).lower()

        if NAMING_ENTITY_EXCEPTIONS.has_key(actual_resource_name_lc):
            actual_resource_name_lc = \
                NAMING_ENTITY_EXCEPTIONS[actual_resource_name_lc]

        headers_method_params_str, headers_map_params_str = \
            HeaderUtils.generate_method_params(link)

        headers_method_params_str = \
            headers_method_params_str + ', ' \
                if headers_method_params_str != '' \
                else headers_method_params_str

        sub_collection_get_template_values = {
          'url':url,
          'parent_resource_name_lc':parent_resource_name_lc.lower(),
          'resource_name_lc':encapsulating_resource.lower(),
          'actual_resource_name_lc':actual_resource_name_lc,
          'headers_method_params_str' : headers_method_params_str,
          'headers_map_params_str' : headers_map_params_str,
          'url_params_id' : UrlUtils.generate_url_identifiers_replacments(
                                  link,
                                  "                            ",
                                  continues=True, is_collection=True
                            ),
          'url_params_name' : UrlUtils.generate_url_identifiers_replacments(
                                  link,
                                  "                    ",
                                  continues=True, is_collection=True
                              ),
          'encapsulating_resource':actual_encapsulating_resource \
                if actual_encapsulating_resource is not None
                                                 else encapsulating_resource,
          'docs':  Documentation.document(
                      link,
                      {
                        'name: string (the name of the entity)': False,
                        'id  : string (the id of the entity)'  : False
                      }
                   )
        }

        return subcollectiongettemplate\
               .generate(sub_collection_get_template_values)

    @staticmethod
    def list(url, link, parent_resource_name_lc, encapsulating_resource,
             actual_resource_name_candidate, KNOWN_WRAPPER_TYPES={},
             NAMING_ENTITY_EXCEPTIONS={}):

        actual_encapsulating_resource = \
                TypeUtil.getValueByKeyOrNone(
                         encapsulating_resource.lower(),
                         KNOWN_WRAPPER_TYPES
                 )

        actual_resource_name_lc = (
                   ParseHelper.getXmlTypeInstance(
                        actual_encapsulating_resource \
                           if actual_encapsulating_resource is not None \
                           else actual_resource_name_candidate.lower()
                   )
        ).lower()

        if NAMING_ENTITY_EXCEPTIONS.has_key(actual_resource_name_lc):
            actual_resource_name_lc = NAMING_ENTITY_EXCEPTIONS[
                                           actual_resource_name_lc
                                      ]

        prms_str, method_params, url_params = \
            ParamUtils.getMethodParamsByUrlParamsMeta(link)

        headers_method_params_str, headers_map_params_str = \
            HeaderUtils.generate_method_params(link)

        combined_method_params = prms_str + \
                                 (', ' if prms_str != '' and headers_method_params_str != '' else '') + \
                                 headers_method_params_str

        method_params_copy = method_params.copy()
        method_params['**kwargs'] = '**kwargs'

        sub_collection_list_template_values = {
           'url':url,
           'parent_resource_name_lc':parent_resource_name_lc.lower(),
           'actual_resource_name_lc':actual_resource_name_lc,
           'combined_method_params':combined_method_params,
           'headers_map_params_str':headers_map_params_str,
           'encapsulating_resource':actual_encapsulating_resource
                    if actual_encapsulating_resource is not None
                    else encapsulating_resource,
           'url_query_params': ParamUtils.toDictStr(
                    url_params.keys(),
                    method_params_copy.keys()
                )
        }



        if prms_str != '' or headers_method_params_str != '':
            sub_collection_list_template_values['docs'] = \
                    Documentation.document(
                       link,
                       {
                            '**kwargs: dict (property based filtering)"': False,
                            'query: string (oVirt engine search dialect query)':False
                        },
                        method_params
            )

            sub_collection_list_template_values['url_params'] = \
                UrlUtils.generate_url_identifiers_replacments(
                      link,
                      "                ",
                      continues=True, is_collection=True
                )

            return subcollectionlistwithparamstemplate\
                   .generate(sub_collection_list_template_values)
        else:
            sub_collection_list_template_values['docs'] = \
                Documentation.document(
                               link,
                               {
                                    '**kwargs: dict (property based filtering)"': False
                               }
                )

            sub_collection_list_template_values['url_params'] = \
                UrlUtils.generate_url_identifiers_replacments(
                              link,
                              "                ",
                              continues=True, is_collection=True
                )

            return subcollectionlisttemplate\
                   .generate(sub_collection_list_template_values)

    @staticmethod
    def add(url, link, body_type, parent_resource_name_lc, encapsulating_entity, KNOWN_WRAPPER_TYPES={}):

        actual_encapsulating_entity = \
            TypeUtil.getValueByKeyOrNone(
                 encapsulating_entity.lower(),
                 KNOWN_WRAPPER_TYPES
             )

        headers_method_params_str, headers_map_params_str = \
            HeaderUtils.generate_method_params(link)

        headers_method_params_str = ', ' + headers_method_params_str \
            if headers_method_params_str != '' \
            else headers_method_params_str

        sub_collection_add_template_values = {
          'resource_to_add':body_type.lower(),
          'url':url,
          'headers_map_params_str':headers_map_params_str,
          'headers_method_params_str':headers_method_params_str,
          'parent_resource_name_lc':parent_resource_name_lc.lower(),
          'encapsulating_entity':actual_encapsulating_entity if actual_encapsulating_entity is not None \
                                                             else encapsulating_entity,
          'docs': Documentation.document(link),
          'url_params':UrlUtils.generate_url_identifiers_replacments(link,
                       "                ",
                       continues=True, is_collection=True)
        }

        return subcollectionaddtemplate\
               .generate(sub_collection_add_template_values)

File: generator/src/main/java/org/ovirt/engine/sdk/generator/rsdl/SubCollection.java
//
// Copyright (c) 2014 Red Hat, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package org.ovirt.engine.sdk.generator.rsdl;

import static org.ovirt.engine.sdk.generator.utils.StringUtils.firstNotNull;

import java.util.LinkedHashMap;
import java.util.Map;

import org.ovirt.engine.sdk.generator.rsdl.templates.SubCollectionAddTemplate;
import org.ovirt.engine.sdk.generator.rsdl.templates.SubCollectionGetTemplate;
import org.ovirt.engine.sdk.generator.rsdl.templates.SubCollectionListTemplate;
import org.ovirt.engine.sdk.generator.rsdl.templates.SubCollectionListWithParamsTemplate;
import org.ovirt.engine.sdk.generator.rsdl.templates.SubCollectionTemplate;
import org.ovirt.engine.sdk.generator.templates.AbstractTemplate;
import org.ovirt.engine.sdk.generator.utils.HeaderUtils;
import org.ovirt.engine.sdk.generator.utils.ParamUtils;
import org.ovirt.engine.sdk.generator.utils.UrlUtils;
import org.ovirt.engine.sdk.generator.xsd.XsdData;
import org.ovirt.engine.sdk.entities.DetailedLink;

public class SubCollection {
    public static String collection(
        String subCollectionName,
        String parentResourceNameLc
    )
    {
        SubCollectionTemplate template = new SubCollectionTemplate();
        template.set("sub_collection_name", subCollectionName);
        template.set("parent_resource_name_lc", parentResourceNameLc.toLowerCase());
        return template.evaluate();
    }

    public static String get(
        String url,
        DetailedLink link,
        String parentResourceNameLc,
        String encapsulatingResource,
        String actualResourceNameCandidate,
        Map<String, String> knownWrapperTypes,
        Map<String, String> namingEntityExceptions
    )
    {
        String actualEncapsulatingResource = knownWrapperTypes.get(encapsulatingResource.toLowerCase());

        String actualResourceNameLc = XsdData.getInstance().getXmlTypeInstance(
            actualEncapsulatingResource != null? actualEncapsulatingResource: actualResourceNameCandidate
        ).toLowerCase();

        if (namingEntityExceptions.containsKey(actualResourceNameLc)) {
            actualResourceNameLc = namingEntityExceptions.get(actualResourceNameLc);
        }

        String[] result = HeaderUtils.generateMethodParams(link);
        String headersMethodParamStr = result[0];
        String headersMapParamsStr = result[1];

        headersMethodParamStr = !headersMethodParamStr.isEmpty()? headersMethodParamStr + ", ": headersMethodParamStr;

        SubCollectionGetTemplate template = new SubCollectionGetTemplate();
        template.set("url", url);
        template.set("parent_resource_name_lc", parentResourceNameLc.toLowerCase());
        template.set("resource_name_lc", encapsulatingResource.toLowerCase());
        template.set("actual_resource_name_lc", actualResourceNameLc);
        template.set("headers_method_params_str", headersMethodParamStr);
        template.set("headers_map_params_str", headersMapParamsStr);
        template.set("url_params_id",
            UrlUtils.generateUrlIdentifiersReplacements(
                link,
                "                            ",
                true,
                true
            )
        );
        template.set("url_params_name",
            UrlUtils.generateUrlIdentifiersReplacements(
                link,
                "                    ",
                true,
                true
            )
        );
        template.set("encapsulating_resource", actualEncapsulatingResource != null? actualEncapsulatingResource: encapsulatingResource);

        Map<String, String> docsParams = new LinkedHashMap<>();
        docsParams.put("id  : string (the id of the entity)", "False");
        docsParams.put("name: string (the name of the entity)", "False");
        String docs = Documentation.document(link, docsParams, new LinkedHashMap<String, String>());
        template.set("docs", docs);

        return template.evaluate();
    }

    public static String list(
        String url,
        DetailedLink link,
        String parentResourceNameLc,
        String encapsulatingResource,
        String actualResourceNameCandidate,
        Map<String, String> knownWrapperTypes,
        Map<String, String> namingEntityExceptions
    )
    {
        String actualEncapsulatingResource = knownWrapperTypes.get(encapsulatingResource.toLowerCase());

        String actualResourceNameLc = XsdData.getInstance().getXmlTypeInstance(
            actualEncapsulatingResource != null? actualEncapsulatingResource: actualResourceNameCandidate.toLowerCase()
        ).toLowerCase();

        if (namingEntityExceptions.containsKey(actualResourceNameLc)) {
            actualResourceNameLc = namingEntityExceptions.get(actualResourceNameLc);
        }

        Object[] result = ParamUtils.getMethodParamsByUrlParamsMeta(link);
        String prmsStr = (String) result[0];
        Map<String, String> methodParams = (Map) result[1];
        Map<String, String> urlParams = (Map) result[2];

        result = HeaderUtils.generateMethodParams(link);
        String headersMethodParamStr = (String) result[0];
        String headersMapParamsStr = (String) result[1];

        String combinedMethodParams =
            prmsStr +
            (!prmsStr.isEmpty() && !headersMethodParamStr.isEmpty()? ", ": "") +
            headersMethodParamStr;

        Map<String, String> methodParamsCopy = new LinkedHashMap<>(methodParams);
        methodParams.put("**kwargs", "**kwargs");

        Map<String, String> values = new LinkedHashMap<>();
        values.put("url", url);
        values.put("parent_resource_name_lc", parentResourceNameLc.toLowerCase());
        values.put("actual_resource_name_lc", actualResourceNameLc);
        values.put("combined_method_params", combinedMethodParams);
        values.put("headers_map_params_str", headersMapParamsStr);
        values.put("encapsulating_resource", firstNotNull(actualEncapsulatingResource, encapsulatingResource));
        values.put("url_query_params", ParamUtils.toDictStr(urlParams.keySet(), methodParamsCopy.keySet()));

        AbstractTemplate template;
        if (!prmsStr.isEmpty() || !headersMethodParamStr.isEmpty()) {
            Map<String, String> customParams = new LinkedHashMap<>();
            customParams.put("**kwargs: dict (property based filtering)\"", "False");
            customParams.put("query: string (oVirt engine search dialect query)", "False");
            String docs = Documentation.document(link, customParams, methodParams);
            values.put("docs", docs);
            values.put("url_params",
                UrlUtils.generateUrlIdentifiersReplacements(
                    link,
                    "                ",
                    true,
                    true
                )
            );
            template = new SubCollectionListWithParamsTemplate();
        }
        else {
            Map<String, String> customParams = new LinkedHashMap<>();
            customParams.put("**kwargs: dict (property based filtering)\"", "False");
            String docs = Documentation.document(link, customParams, new LinkedHashMap<String, String>());
            values.put("docs", docs);
            values.put("url_params",
                UrlUtils.generateUrlIdentifiersReplacements(
                    link,
                    "                ",
                    true,
                    true
                )
            );
            template = new SubCollectionListTemplate();
        }

        template.set(values);
        return template.evaluate();
    }

    public static String add(
        String url,
        DetailedLink link,
        String bodyType,
        String parentResourceNameLc,
        String encapsulatingEntity,
        Map<String, String> knownWrapperTypes
    )
    {
        String actualEncapsulatingEntity = knownWrapperTypes.get(encapsulatingEntity.toLowerCase());

        String[] result = HeaderUtils.generateMethodParams(link);
        String headersMethodParamStr = result[0];
        String headersMapParamsStr = result[1];

        headersMethodParamStr = !headersMethodParamStr.isEmpty()?
            ", " + headersMethodParamStr:
            headersMethodParamStr;

        SubCollectionAddTemplate template = new SubCollectionAddTemplate();
        template.set("resource_to_add", bodyType.toLowerCase());
        template.set("url", url);
        template.set("headers_map_params_str", headersMapParamsStr);
        template.set("headers_method_params_str", headersMethodParamStr);
        template.set("parent_resource_name_lc", parentResourceNameLc.toLowerCase());
        template.set("encapsulating_entity", firstNotNull(actualEncapsulatingEntity, encapsulatingEntity));
        template.set("docs", Documentation.document(link));
        template.set("url_params",
            UrlUtils.generateUrlIdentifiersReplacements(
                link,
                "                ",
                true,
                true
            )
        );

        return template.evaluate();
    }
}
